// @progress/kendo-angular-editor/dist/npm/tools/shared/editor-command-dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var command_icons_1 = require("../../config/command-icons");
/**
 * @hidden
 */
var EditorCommandDialog = /** @class */ (function () {
    function EditorCommandDialog(dialog, button, editor) {
        this.dialog = dialog;
        this.button = button;
        this.editor = editor;
    }
    EditorCommandDialog.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.button.icon) {
            this.button.icon = command_icons_1.commandIcons[this.dialog];
        }
        this.subs = this.button.click.subscribe(function () {
            _this.editor.openDialog(_this.dialog);
        });
        this.subs.add(this.editor.stateChange.subscribe(this.onStateChange.bind(this)));
    };
    EditorCommandDialog.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    /**
     * @hidden
     */
    // tslint:disable-next-line
    EditorCommandDialog.prototype.onStateChange = function (_toolBarState) { };
    return EditorCommandDialog;
}());
exports.EditorCommandDialog = EditorCommandDialog;
