// @progress/kendo-angular-editor/dist/es/tools/shared/editor-command-button.js
import { commandIcons } from '../../config/command-icons';
/**
 * @hidden
 */
var EditorCommandButton = /** @class */ (function () {
    function EditorCommandButton(command, button, editor) {
        this.command = command;
        this.button = button;
        this.editor = editor;
    }
    EditorCommandButton.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.button.icon) {
            this.button.icon = commandIcons[this.command];
        }
        this.subs = this.button.click.subscribe(function () {
            _this.editor.exec(_this.command);
            _this.editor.focus();
        });
        this.subs.add(this.editor.stateChange.subscribe(this.onStateChange.bind(this)));
    };
    EditorCommandButton.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    // tslint:disable-next-line:typedef
    EditorCommandButton.prototype.onStateChange = function (toolBarState) {
        this.button.selected = toolBarState[this.command];
    };
    return EditorCommandButton;
}());
export { EditorCommandButton };
